controller.py

```diff
@@ -192,9 +193,63 @@ def canonicalize_author_name(self):
         return make_response(author_name, HTTP_OK, {"Content-Type": "text/plain"})
 
 
-class CatalogController(object):
+class ISBNEntryMixin(object):
+
+    def make_opds_entry_from_metadata_lookups(self, identifier):
+        """This identifier cannot be turned into a presentation-ready Work,
+        but maybe we can make an OPDS entry based on metadata lookups.
+        """
+
+        # We can only create an OPDS entry if all the lookups have
+        # in fact been done.
+        metadata_sources = DataSource.metadata_sources_for(
+            self._db, identifier
+        )
+        q = self._db.query(CoverageRecord).filter(
+                CoverageRecord.identifier==identifier
+        ).filter(
+            CoverageRecord.data_source_id.in_([x.id for x in metadata_sources]),
+            CoverageRecord.status==CoverageRecord.SUCCESS
+        )
+
+        coverage_records = q.all()
+        unaccounted_for = set(metadata_sources)
+        for r in coverage_records:
+            if r.data_source in unaccounted_for:
+                unaccounted_for.remove(r.data_source)
+
+        if unaccounted_for:
+            # At least one metadata lookup has not successfully
+            # completed.
+            names = [x.name for x in unaccounted_for]
+            self.log.info(
+                "Cannot build metadata-based OPDS feed for %r: missing coverage records for %s",
+                identifier,
+                ", ".join(names)
+            )
+            return None
```

Comment:
Instead of directly adding the resulting message or entry to the precomposed entries, it just returns the message/entry itself. Or None, if there's no entry to be had.